### DISCUSSION TITLE
Get altitude reading from landscape

### DISCUSSION THREAD

**Discussion by djdylex** [microsoft/AirSim#3508] [Support Q&A]
Hi, 

I am running airsim on a custom map with a landscape that uses a heightmap. The drone reads its altitute as its z value, not the distance from itself to the ground. This causes the drone to collide with the ground after the ground slopes up.

Is there any way I can have the altitude be read from the ground rather than the absolute z component so that when i call " client.moveByVelocityZAsync" it will stay at the specified altitude above the landscape?

**Comment by rajat2004** [comment]
It's not exactly clear to me what you're trying to achieve; from what I understood, you want the drone to always be, let's say, 40m above the ground, regardless of the terrain beneath it, such as hills or plains, etc. Something like https://ardupilot.org/copter/docs/terrain-following.html?

As mentioned in the doc, this is generally accomplished by storing the terrain data on the controller itself, which itself is sent to the vehicle by the GCS. The other option is to use a downward-facing sensor for altitude measurement. This can be done using a Distance Sensor; see https://microsoft.github.io/AirSim/sensors/#distance-sensor. Using ArduPilot or PX4, it can be set up to work automatically.

Using just the Python APIs with SimpleFlight, you could periodically get the current height (distance from the ground) and call the API to change the altitude as required. Just add the desired correction (if its 35m above the ground, add 5m to the current Z value, and call the API). There is also a recent PR https://github.com/microsoft/AirSim/pull/3475 which will allow values to be set relative to the body frame.

Hope this helps!

**Reply by djdylex** [reply]
That clears it up a lot thankyou! Not sure why any of this didn't come up in my searches
